### DISCUSSION TITLE
Have you considered adding the possibility of specifying custom data sources?

### DISCUSSION THREAD

**Discussion by ionut-botizan** [keystonejs/keystone#6614] [Event Q&A Q3 2021]
Hi,

What I mean by that is, instead of generating a Prisma schema/model for some entity, let developers specify a custom data source, by allowing them to implement custom functions that perform the CRUD operations (for example, by querying existing APIs that already manage some part of the content).
I do realize this might be a bit problematic when it comes to defining relations so let's just say that there will be no relationships possible with these external entities (or there would be, but in a more limited form).
Is this something you would consider doing in the medium to far future? Or, would you be open to a pull request that implements this?

Regards,
Ionuț

**Comment by timleslie** [comment]
Hi @ionut-botizan, this is already supported via the GraphQL schema extension API. Basically, if you have an external datasource you need to query, you can define your own GraphQL queries and mutations which interact with this datasource, and have the resolvers for those operations do whatever you need 👍
